File: 97-interleaving-string/interleaving-string.java
class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        if (s1.length() + s2.length() != s3.length()) return false;
        Boolean[][] memo = new Boolean[s1.length() + 1][s2.length() + 1];
        return fun(0, 0, s1, s2, s3, memo);
    }

    public boolean fun(int i1, int i2, String s1, String s2, String s3, Boolean[][] memo) {
        if (i1 + i2 == s3.length()) return true;

        if (memo[i1][i2] != null) return memo[i1][i2];

        boolean pickLeft = false, pickRight = false;

        if (i1 < s1.length() && s1.charAt(i1) == s3.charAt(i1 + i2)) {
            pickLeft = fun(i1 + 1, i2, s1, s2, s3, memo);
        }

        if (i2 < s2.length() && s2.charAt(i2) == s3.charAt(i1 + i2)) {
            pickRight = fun(i1, i2 + 1, s1, s2, s3, memo);
        }

        return memo[i1][i2] = pickLeft || pickRight;
    }
}

File: 97-interleaving-string/interleaving-string.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1) + len(s2) != len(s3):
            return False
        dp = {}
        def fun(i1,i2):
            i = i1 + i2
            if (i1,i2) in dp:
                return dp[(i1,i2)]
            if i == len(s3):
                return True
            if i1 < len(s1) and s1[i1] == s3[i]:
                if fun(i1+1,i2):
                    dp[(i1+1,i2)] = True
                    return True
            if i2 < len(s2) and s2[i2] == s3[i]:
                if fun(i1,i2+1):
                    dp[(i1,i2+1)] = True
                    return True
            dp[(i1,i2)] = False
            return False
        return fun(0,0)
